ironfish/src/rpc/routes/chain/importChain.ts

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import bufio from 'bufio'
+import * as yup from 'yup'
+import { Assert } from '../../../assert'
+import { readBlock } from '../../../network/utils/block'
+import { IronfishNode } from '../../../node'
+import { Block } from '../../../primitives'
+import { ApiNamespace, router } from '../router'
+
+export type ImportSnapshotRequest =
+  | {
+      chunkStart?: number | null
+      chunkEnd?: number | null
+      blocks?: Buffer | null
+    }
+  | undefined
+
+export type ImportSnapshotResponse = {
+  headSeq: number
+}
+
+export const ImportSnapshotRequestSchema: yup.ObjectSchema<ImportSnapshotRequest> = yup
+  .object({
+    chunkStart: yup.number().nullable().optional(),
+    chunkEnd: yup.number().nullable().optional(),
+    blocks: yup.mixed<Buffer>().nullable().optional(),
+  })
+  .optional()
+
+export const ImportSnapshotSchema: yup.ObjectSchema<ImportSnapshotResponse> = yup
+  .object({
+    headSeq: yup.number().defined(),
+  })
+  .defined()
+
+router.register<typeof ImportSnapshotRequestSchema, ImportSnapshotResponse>(
+  `${ApiNamespace.chain}/importSnapshot`,
+  ImportSnapshotRequestSchema,
+  async (request, node): Promise<void> => {
+    Assert.isNotNull(node.chain.head, 'head')
+    Assert.isNotNull(node.chain.latest, 'latest')
+
+    if (!request.data?.blocks) {
+      request.end({ headSeq: node.chain.head.sequence })
+    }
+
+    const blocks = request.data?.blocks
+    if (blocks) {
+      const reader = bufio.read(blocks, true)
+      const deserializedBlocks = deserializeChunk(node, reader)
+
+      for (const block of deserializedBlocks) {
+        const present = await node.chain.hasBlock(block.header.hash)
+        if (!present) {
+          const result = await node.chain.addBlock(block)
+          if (!result.isAdded && result.reason) {
+            node.logger.debug(
+              `Could not add block ${block.header.sequence} from snapshot, reason: ${result.reason}`,
+            )
+          }
+        }
+      }
+
+      request.end({ headSeq: node.chain.head.sequence })
+    }
+  },
+)
+
+function deserializeChunk(node: IronfishNode, reader: bufio.BufferReader): Block[] {
+  const blocks: Block[] = []
+  const blocksLength = reader.readU64()
+  for (let i = 0; i < blocksLength; i++) {
+    const blockBuffer = reader.readVarBytes()
+    const blockReader = bufio.read(blockBuffer, true)
```

Comment:
Is it necessary to read this into a separate Buffer first? I think you might just be able to call `readBlock(reader)`

Reply:
In the export step, each of the blocks is written as a var byte string, so a reader is instantiated with the contents of that variable length string. I did clean it up a bit, however.

Reply:
Ah, I should have code reviewed that better in the export PR. This is fine